src/transforms/kubernetes/mod.rs

```diff
@@ -1 +1,552 @@
+            let error = metadata_client.run().await;
+            error!(message = "Stopped updating Pod metadata.", reason = %error);
+        });
+
+        // Construct transform
+        Ok(Box::new(KubernetesPodMetadata { metadata: reader }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kubernetes_pod_metadata"
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Failed building WatchClient: {}", source))]
+    WatchClientBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display("Failed building watch stream: {}", source))]
+    WatchStreamBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display(
+        "Missing environment variable {:?} containing node name `spec.nodeName`.",
+        env
+    ))]
+    MissingNodeName { env: &'static str },
+    #[snafu(display("Unknown metadata fields: {:?}", fields))]
+    UnknownFields { fields: Vec<String> },
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
+    /// (key of data to be deleted, can be deleted after this point in time)
+    delete_queue: VecDeque<(Bytes, Instant)>,
+    client: WatchClient,
+}
+
+impl MetadataClient {
+    fn new(
+        trans_config: &KubePodMetadata,
+        metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
+        client: WatchClient,
+    ) -> Result<Self, BuildError> {
+        // Select Pod metadata fields that need to be extracted from
+        // Pod and then added to Events.
+        let mut add_fields = trans_config.fields.clone();
+        add_fields.sort();
+        add_fields.dedup();
+
+        let mut fields = Vec::new();
+        let mut unknown = Vec::new();
+        let mut all_fields = all_fields();
+
+        for field in add_fields {
+            if let Some(fun) = all_fields.remove(field.as_str()) {
+                fields.push(fun);
+            } else {
+                unknown.push(field.to_owned());
+            }
+        }
+
+        if unknown.is_empty() {
+            Ok(Self {
+                fields,
+                metadata,
+                client,
+                delete_queue: VecDeque::new(),
+            })
+        } else {
+            Err(BuildError::UnknownFields { fields: unknown })
+        }
+    }
+
+    /// Listens for pod metadata changes and updates metadata map.
+    async fn run(mut self) -> BuildError {
+        let mut version = None;
+        let mut error = None;
+        loop {
+            // Build watcher stream
+            let mut watcher = match self.client.watch_metadata(version.clone(), error.take()) {
```

Comment:
It would remove only that one match, but then it wouldn't be as clear that this function ends only when it errors. And on first glance Result would mislead that it can end in other ways.

Reply:
Fair, but I also find it a bit odd that this fn returns an error without using a Result. I think to me that makes this much more clear to understand.

Reply:
True, I'll change it.